http/server.ts

```diff
@@ -1,111 +1,117 @@
-    return;
-  }
-  env.reqQueue.push(req); // push req to queue
-  env.serveDeferred.resolve(); // signal while loop to process it
+export interface ServerResponse {
+  status?: number;
+  headers?: Headers;
+  body?: Uint8Array | Reader;
 }
 
-export async function* serve(addr: string) {
+export async function* serve(
+  addr: string,
+  cancel: Deferred<void> = defer<void>()
+): AsyncIterableIterator<ServerRequest> {
   const listener = listen("tcp", addr);
-  const env: ServeEnv = {
-    reqQueue: [], // in case multiple promises are ready
-    serveDeferred: deferred()
-  };
-
-  // Routine that keeps calling accept
-  const acceptRoutine = () => {
-    const handleConn = (conn: Conn) => {
-      serveConn(env, conn); // don't block
-      scheduleAccept(); // schedule next accept
-    };
-    const scheduleAccept = () => {
-      listener.accept().then(handleConn);
-    };
-    scheduleAccept();
-  };
-
-  acceptRoutine();
-
-  // Loop hack to allow yield (yield won't work in callbacks)
   while (true) {
-    await env.serveDeferred.promise;
-    env.serveDeferred = deferred(); // use a new deferred
-    let queueToProcess = env.reqQueue;
-    env.reqQueue = [];
-    for (const result of queueToProcess) {
-      yield result;
-      // Continue read more from conn when user is done with the current req
-      // Moving this here makes it easier to manage
-      serveConn(env, result.conn, result.r);
+    const raced = await Promise.race<Conn | void>([
+      listener.accept(),
+      cancel.promise
+    ]);
+    if (!raced) {
+      break;
     }
+    yield readRequest(raced, async response => {
+      await writeResponse(raced, response);
+    });
   }
   listener.close();
 }
 
-export async function listenAndServe(
-  addr: string,
-  handler: (req: ServerRequest) => void
-) {
+export async function listenAndServe(addr: string, handler: HttpHandler) {
   const server = serve(addr);
 
-  for await (const request of server) {
-    await handler(request);
+  for await (const req of server) {
+    await handler(req);
   }
 }
 
-export interface Response {
-  status?: number;
-  headers?: Headers;
-  body?: Uint8Array | Reader;
+export type HttpServerHandler = HttpHandler;
+
+export interface HttpServer {
+  handle(pattern: string, handler: HttpHandler);
+
+  listen(addr: string, cancel?: Deferred<void>): Promise<void>;
+}
+
+export function createServer(): HttpServer {
+  return new HttpServerImpl();
+}
+
+class HttpServerImpl implements HttpServer {
+  private handlers: {
+    [key: string]: {
+      pattern: string;
+      regexp: RegExp;
+      keys?: Key[];
+      handler: HttpHandler;
+    };
+  } = Object.create(null);
+
+  handle(pattern: string, handler: HttpHandler) {
+    const keys = [];
+    const regexp = pathToRegexp(pattern, keys);
+    this.handlers[pattern] = {
+      pattern,
+      regexp,
+      keys,
+      handler
+    };
+  }
+
+  async listen(addr: string, cancel: Deferred<void> = defer<void>()) {
```

Comment:
I see 2 edge cases in this method:
- developer not calling `request.respond` in `handler`
- no URL match
I think it will hang the request... I'd opt to throw error if `respond` is not called and return 404 response in respective situations

Reply:
Exactly. I will add fallbacks.